lib/bats-core/test_functions.bash

```diff
@@ -29,7 +29,37 @@ load() {
   source "${file}"
 }
 
+#########
+#  die  #  Abort with helpful and visible message
+#########
+die() {
+  printf "#/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv\n"  >&2
+  printf "#| FAIL: %s\n" "$1"                                      >&2
+  shift
+  # Any more?
+  for line in "$@"; do
+    printf "#|    > %s\n" "$line"                                  >&2
+  done
+
+  printf "#\\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n" >&2
+  false
+}
+
 run() {
+  local expected_rc=
+
+  # Possible first arguments: '!' = any nonzero, '=n' = exactly n
+  if [[ "$1" = '!' ]]; then
+    expected_rc=-1
+    shift
+  elif [[ "$1" =~ ^=([0-9]|[1-9][0-9]|[12][0-9][0-9])$ ]]; then
+    expected_rc=${1#=}
+    shift
+  fi
+
+  # stdout is emitted only on error; this echo is to help debug test failures
+  printf "${BATS_PS1}%s\n" "$*"
```

Comment:
I understand that this should print the command under `run` even when there is no failure. In the failure case I find this to duplication to be counter-productive. In the case of a later failure (not in this run) I would reserve this behavior for an eventual trace functionality.

Reply:
May I push back on this? I can't count the number of times this has saved us in analyzing test failures: being able to see earlier commands run, and their output, has been invaluable when trying to understand a sequence of events leading to failures. Put another way, the root cause of a failure is sometimes buried in a command run some time before.

Reply:
Yes, I experienced something like this myself. However, one of the core tenets of `run` is being silent to the log by default. As indicated above, I have absolutely nothing against allowing a switch like `--trace` that would make run more verbose if the user chooses to.

Reply:
>one of the core tenets of run is being silent to the log by default

Is that a hard requirement, or is it even documented? I've been struggling with this comment, because without verbose `run` I don't see any point to this PR. But I think maybe I haven't described how it works well enough. This usage is absolutely silent on everything except failure; and on failure, it only shows `run`s for the failed test - nothing else. Given a test setup with lots of tiny tests, this does not add clutter. (If, OTOH, you have one giant `@test` that runs hundreds of commands, it could be harder to read). 

To put it another way, here's a [recent example](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/5159711932678144/html/sys-podman-fedora-33-rootless-host.log.html) of a failure we had today. In this case the failure message shows container `0e7a2d3`, but that would be utterly useless without the earlier context in which we see `0e7a2d3` as the output of the second `podman run`. We see failures like these often, typically race conditions, where every little bit of context is critical for identifying the cause. As you can see, only the failing test shows `run` output; the other 100+ passing/skipped tests are silent. Does that help alleviate your concerns?

Reply:
> Is that a hard requirement, or is it even documented?

I don't think it is documented and I'd say it is not a hard requirement per se but you have to remember that there are loads and loads of tests out there which rely on the current default behavior. 

> This usage is absolutely silent on everything except failure

Yes, that is the default behavoir of bats. However, there are cases like this:

```
run succeeding_command_with_enormous_output
assert_output ...

run failing_command_with_output
assert_output ...
```

Where you'd have to wade through lots and lots of uninteresting output. The current implementation makes all users pay for this. All I am asking for is a switch that allows users to decide for themselves if they want that much output or not. This is what I was alluding to with the (not yet implemented) `--trace` switch. You could start the whole bats run with that flag or we should allow for more finegrained control, e.g. via 

```
BATS_TRACE=only-commands # print only commands in run here
run command 1
...
run command n
BATS_TRACE=output # print commands and stdout/stderr here
run command n+1
...
unset BATS_TRACE # print neither command nor output in the following
run command 
```

The unset `BATS_TRACE` case resembles the current default and should remain the default.

I have seen several tests where the amount of output of a single command is too much and there are several in the test.